File: leetcode-solutions/ruby/15. 3Sum.rb
# https://leetcode.com/problems/3sum/
# TODO: lime limite exceed submission

# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
  res = []
  comb = []
  combination(res, comb, nums.sort, 0)
  res
end

def combination(res, comb, nums, s_index)
  return if comb.sum > 0
  
  if comb.size == 3 && comb.sum.zero?
      res << comb.map{|n| n}
      return
  end
  
  if comb.size == 3
      return 
  end
  
  s_index.upto(nums.size-1) do |n|
      next if (n > s_index  && nums[n] == nums[n-1]) || nums[n] > 0
      comb << nums[n]
      combination(res, comb, nums, n+1)
      comb.pop
  end
end

File: leetcode-solutions/python/15. 3Sum.py
# https://leetcode.com/problems/3sum/

'''
use two pointer technique to find the two sum calculation and use that to calcumate the 3 sum
'''
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        length = len(nums)
        
        def twoSum(elm, s, e = len(nums) - 1):
            while s < e:
                if nums[s] + nums[e] + elm > 0:
                    e -= 1
                elif nums[s] + nums[e] + elm < 0:
                    s += 1
                else:
                   # sum of two is expected elm                     
                    comb = [elm, nums[s], nums[e]]
                    if not comb in ans:
                        ans.append(comb)
                    s += 1
                    e -= 1
        
        for i in range(length - 2):
            twoSum(nums[i], i + 1)
        return ans
            
                
        
            
            